### PR TITLE
CI friendly fast mode for try state checks

### LINKED ISSUE #234: Staking `try_state` times out
CI timeout after 120 min for try-state: [Westend](https://gitlab.parity.io/parity/mirrors/polkadot/-/jobs/2214152) and [Polkadot](https://gitlab.parity.io/parity/mirrors/polkadot/-/jobs/2214152) tested [here](https://github.com/paritytech/polkadot/pull/6491) (first commit).  
Not sure which pallet it is exactly, but AFAIK something from staking.

### PR DESCRIPTION
**Pull request by Ank4n**
Reviving https://github.com/paritytech/substrate/pull/13286
Should resolve https://github.com/paritytech/polkadot-sdk/issues/234

Adds a new fast mode to `TryStateSelect` which skips slow running state checks.

Adds two new options `[fast-all | fast-try-state]` to `on-runtime-upgrade --checks`.

Example usage:

```bash
 ./polkadot \
     try-runtime \
     --runtime runtime-try-runtime.wasm \
     -lruntime=debug \
     on-runtime-upgrade \
     --checks=fast-all \
     live \
     --uri ws://localhost:9944
 ```

### CODE REVIEW

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2386987

File: frame/nomination-pools/src/lib.rs [hunk 1]
```diff
@@ -1172,6 +1172,9 @@ pub mod pallet {
 	use frame_support::traits::StorageVersion;
 	use frame_system::{ensure_signed, pallet_prelude::*};
 
+	#[cfg(feature = "try-runtime")]
+	use frame_support::traits::TryStateSelect;
```
Comment (kianenigma) [inline frame/nomination-pools/src/lib.rs:1176]: I usually use the full path to avoid feature gated imports.

File: frame/executive/src/lib.rs [hunk 2]
```diff
@@ -353,7 +353,7 @@ where
 			let _guard = frame_support::StorageNoopGuard::default();
 			<AllPalletsWithSystem as frame_support::traits::TryState<System::BlockNumber>>::try_state(
 				frame_system::Pallet::<System>::block_number(),
-				frame_try_runtime::TryStateSelect::All,
+				frame_try_runtime::TryStateSelect::Fast,
```
[the remark this answers is not in the document]
Comment (kianenigma) [reply frame/executive/src/lib.rs:356]: So basically @Ank4n had asked: 

> Should we add another option in UpgradeCheckSelect to choose try-state-fast checks?

I'd say yes.

File: frame/nomination-pools/src/lib.rs (hunk 1 above)
Comment (Ank4n) [reply frame/nomination-pools/src/lib.rs:1176]: The mod `try_runtime` is private.

File: frame/staking/src/pallet/mod.rs [hunk 3]
```diff
@@ -825,8 +827,8 @@ pub mod pallet {
 		}
 
 		#[cfg(feature = "try-runtime")]
-		fn try_state(n: BlockNumberFor<T>) -> Result<(), &'static str> {
-			Self::do_try_state(n)
+		fn try_state(n: BlockNumberFor<T>, select: TryStateSelect) -> Result<(), &'static str> {
+			Self::do_try_state(n, select == TryStateSelect::Fast)
```
[the remark this answers is not in the document]
Comment (Ank4n) [reply frame/staking/src/pallet/mod.rs:831]: Yeah we could. I like this better though since the `do_try_state` does not need to know anything except whether its fast mode or not.

File: frame/bags-list/src/lib.rs [hunk 4]
```diff
@@ -98,6 +98,9 @@ pub mod pallet {
 	use frame_support::pallet_prelude::*;
 	use frame_system::pallet_prelude::*;
 
+	#[cfg(feature = "try-runtime")]
+	use frame_support::traits::TryStateSelect;
```
Comment (kianenigma) [inline frame/bags-list/src/lib.rs:102]: I am generally a fan of using full paths instead of feature gated imports.

File: frame/bags-list/src/lib.rs [hunk 5]
```diff
@@ -268,7 +271,7 @@ pub mod pallet {
 		}
 
 		#[cfg(feature = "try-runtime")]
-		fn try_state(_: BlockNumberFor<T>) -> Result<(), &'static str> {
+		fn try_state(_: BlockNumberFor<T>, _: TryStateSelect) -> Result<(), &'static str> {
```
Comment (kianenigma) [inline frame/bags-list/src/lib.rs:274]:
```suggestion
		fn try_state(_: BlockNumberFor<T>, _: frame_support::traits::TryStateSelect) -> Result<(), &'static str> {
```

eg

File: frame/nomination-pools/src/lib.rs (hunk 1 above)
Comment (kianenigma) [reply frame/nomination-pools/src/lib.rs:1176]: Everything should be re-exported from `frame-support::traits`

File: frame/bags-list/src/lib.rs (hunk 4 above)
Comment (kianenigma) [reply frame/bags-list/src/lib.rs:102]: Seems like we talked about it before :D https://github.com/paritytech/substrate/pull/13382/files#r1105799576

File: frame/nomination-pools/src/lib.rs [hunk 6]
```diff
@@ -2122,7 +2125,7 @@ pub mod pallet {
 	#[pallet::hooks]
 	impl<T: Config> Hooks<BlockNumberFor<T>> for Pallet<T> {
 		#[cfg(feature = "try-runtime")]
-		fn try_state(_n: BlockNumberFor<T>) -> Result<(), &'static str> {
+		fn try_state(_n: BlockNumberFor<T>, _: TryStateSelect) -> Result<(), &'static str> {
```
Comment (kianenigma) [inline frame/nomination-pools/src/lib.rs:2128]:
```suggestion
		fn try_state(_n: BlockNumberFor<T>, _: frame_support::traits::TryStateSelect) -> Result<(), &'static str> {
```

Okay this should work no?

File: frame/support/src/traits/try_runtime.rs [hunk 7]
```diff
@@ -34,6 +34,10 @@ pub enum Select {
 	///
 	/// Pallet names are obtained from [`super::PalletInfoAccess`].
 	Only(Vec<Vec<u8>>),
+	/// Run only fast running tests.
```
Comment (kianenigma) [inline frame/support/src/traits/try_runtime.rs:37]: This will run all pallets, but in fast mode right? Or else I am not sure how it work with that regards.

File: frame/support/src/traits/hooks.rs [hunk 8]
```diff
@@ -272,7 +274,7 @@ pub trait Hooks<BlockNumber> {
 	///
 	/// This hook should not alter any storage.
 	#[cfg(feature = "try-runtime")]
-	fn try_state(_n: BlockNumber) -> Result<(), &'static str> {
+	fn try_state(_n: BlockNumber, _s: TryStateSelect) -> Result<(), &'static str> {
```
Comment (kianenigma) [inline frame/support/src/traits/hooks.rs:277]: new argument needs to be documented.

**Review by kianenigma — COMMENTED** [review]
Needs a bit more work, looking forward to it + finally enabling these test in CI.

This is also relevant for #13013, probably decoding the entire state is not something that we want to do in CI. Not sure.

All in all, my priority is to make these test and the CI work. @Ank4n fwiw I would happily approve a temp solution that simply reduces the `try_state` of all pallets, including staking, to be something sensible (perhaps feature gated by `env!("CI_EXEC")` as a quick hack) and then we can incrementally introduce them back, or think about how to do it.

In other words, in the case of this PR, getting something out there fast is more important than perfection.

Please check with @ggwpez about enabling CI checks as a lot of the recent CI checks have been his curtesy.

**Comment by kianenigma** [comment]
Updates here?

**Comment by Ank4n** [comment]
> > Updates here?
> 
> Repeat 😁

I am working on refreshing this PR.

File: frame/support/src/traits/try_runtime.rs (hunk 7 above)
Comment (Ank4n) [reply frame/support/src/traits/try_runtime.rs]: Yes, it is same as `All` but the pallets can choose to ignore some tests in the fast mode.

File: utils/frame/try-runtime/cli/src/commands/on_runtime_upgrade.rs [hunk 9]
```diff
@@ -34,8 +34,10 @@ pub struct OnRuntimeUpgradeCmd {
 	///
 	/// - `none`: Perform no checks (default when the arg is not present).
 	/// - `all`: Perform all checks (default when the arg is present).
+	/// - `fast-all`: Perform fast running state, pre- and post-upgrade checks.
 	/// - `pre-and-post`: Perform pre- and post-upgrade checks.
 	/// - `try-state`: Perform the try-state checks.
+	/// - `fast-try-state`: Perform fast running state checks.
```
Comment (Ank4n) [inline utils/frame/try-runtime/cli/src/commands/on_runtime_upgrade.rs:40]: May be a better approach would be to add another option to the `OnRuntimeUpgrade` command, such as `mode: [fast | normal]`.

File: frame/executive/src/lib.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (Ank4n) [reply frame/executive/src/lib.rs]: Selecting `TryState` won't run pre and post checks though.

**Review by liamaharon — COMMENTED** [review]
This is a cool idea, but I'd like to raise the possibility that it may be better to completely turn off try-state checks in the CI than to have them run partially.

- Run partially, the green CI check loses its meaning. It no longer would signify that try-state hooks are passing, rather just that they *may* be passing. To be sure they're actually passing, they would need to be run again somewhere else anyway, so there seems to be little value in running them in the CI. 
    - At most, they could alert the dev to some failing hooks. But, the dev would still need to run the full set of try-state hooks somewhere else anyway.
- The green check may provide a false sense of security to devs who're unaware that the full set of hooks are not run in the CI.
- The feature comes with a cost: it adds extra configuration to the cli, and introduces cognitive overhead for developers working on the try-state hooks needing to decide whether a check should be 'fast mode' or not. 

I'm very open to being shown why the cost is justified here, just want to make sure this has been considered

**Comment by Ank4n** [comment]
@liamaharon Thanks for raising those points. Its still all open to discussion but I will write how I was thinking about this.

> This is a cool idea, but I'd like to raise the possibility that it may be better to completely turn off try-state checks in the CI than to have them run partially.

Currently the try-state checks are disabled on CI (it only runs pre- and post-upgrade checks). So in a way what you are suggesting is what is happening currently.

> * Run partially, the green CI check loses its meaning. It no longer would signify that try-state hooks are passing, rather just that they _may_ be passing. To be sure they're actually passing, they would need to be run again somewhere else anyway, so there seems to be little value in running them in the CI.

Another way to think about it is, this would allow every pallet to write very thorough try state tests. Some of these tests may take really long time (probably hours) and it would be great to run all of them on CI but we also don't want CI to be stuck for hours. The pallet developer still believes other tests would ensure 99.9% scenarios are covered and this slow running test is something they run occasionally outside CI to ensure all storage items are consistent to the expectations.

To give a more concrete example (which is actually the reason why we wanted to introduce the fast mode) we can look at [this try-state](https://github.com/paritytech/substrate/blob/master/frame/staking/src/pallet/impls.rs#L1778) check in staking pallet. It iterates over all nominators (~ 44k) and then for each nominator, it fetches all active validators (~ 300), gets their exposure, iterates over all of its stakers (could be upto 512) and checks if the nominator (in the first loop) is only present once in a validator's exposure. This is cubic time complexity and takes more than 2 hours to run currently. 

Also, most of the times, a failing try-state checks may not be introduced by the current PR but through a sneaky bug that  we only found out about later. I think we might even want to keep these tests to `allow_failure` and these are just indications to investigate and not blockers to a PR. 

To summarise, I think what this PR is trying to do is find a middle ground where we can run try state checks in the CI that would cover most of the inconsistent state scenario but still enable pallet developer write some more thorough but expensive tests that they can run outside CI if they want to. Its also important to emphasise try-state check should only be our 3rd or 4th line of defence.


>   * At most, they could alert the dev to some failing hooks. But, the dev would still need to run the full set of try-state hooks somewhere else anyway.
> * The green check may provide a false sense of security to devs who're unaware that the full set of hooks are not run in the CI.
> * The feature comes with a cost: it adds extra configuration to the cli, and introduces cognitive overhead for developers working on the try-state hooks needing to decide whether a check should be 'fast mode' or not.

I agree we should look to improve this, make it more intuitive and better documented. By default every test should be marked as fast test unless we notice they are taking really long to run in the CI. I do think though that running 99% of the try state checks on CI and ignoring one is better alternative than not running anything on CI. If there are complex changes to a pallet logic, a developer should always run full suite (may be we can have a bot command that runs the full suite on demand) and even possibly add more state checks.

**Review by liamaharon — APPROVED** [review]

File: frame/support/src/traits/hooks.rs [hunk 10]
```diff
@@ -306,9 +308,12 @@ pub trait Hooks<BlockNumber> {
 	/// It should focus on certain checks to ensure that the state is sensible. This is never
 	/// executed in a consensus code-path, therefore it can consume as much weight as it needs.
 	///
+	/// Takes the block number and `TryStateSelect`as a parameter. The `TryStateSelect` is used to
```
Comment (gpestana) [inline frame/support/src/traits/hooks.rs:311]:
```suggestion
	/// Takes the block number and `TryStateSelect` as a parameter. The `TryStateSelect` is used to
```

File: frame/staking/src/mock.rs [hunk 11]
```diff
@@ -550,7 +550,7 @@ impl ExtBuilder {
 		let mut ext = self.build();
 		ext.execute_with(test);
 		ext.execute_with(|| {
-			Staking::do_try_state(System::block_number()).unwrap();
+			Staking::do_try_state(System::block_number(), false).unwrap();
```
Comment (gpestana) [inline frame/staking/src/mock.rs:553]: maybe it would be more clear for the reader to pass one of [`TryStateSelect::Fast`, `TryStateSelect::All`] here instead of bool.

**Review by gpestana — APPROVED** [review]
Neat!

File: frame/collective/src/lib.rs [hunk 12]
```diff
@@ -349,7 +349,10 @@ pub mod pallet {
 	#[pallet::hooks]
 	impl<T: Config<I>, I: 'static> Hooks<BlockNumberFor<T>> for Pallet<T, I> {
 		#[cfg(feature = "try-runtime")]
-		fn try_state(_n: BlockNumberFor<T>) -> Result<(), TryRuntimeError> {
+		fn try_state(
+			_n: BlockNumberFor<T>,
+			_s: frame_support::traits::TryStateSelect,
```
Comment (kianenigma) [inline frame/collective/src/lib.rs:354]: This API is unfortunately not correct. 

You are right to pass in _something_ into the hook that helps it understand if it is fast or not, but `TryStateSelect` is not the right type here. 

`TryStateSelect` is meant to identify which pallets to execute, not how much time they should each consume. It is only interpreted by the `Executive` and should not be exposed to the end user here at all. 

What we instead want is a `TryStateSpeed` which you can for now assume to `bool` or `enum TryStateSpeed { Slow, Mid, Fast }`. 

Then, you are capable of selecting which pallets to run, and at what speed. 

As it stands now, I see this flaw as welll: 

A pallet could possibly see `RoundRobin(7)` as its try-state select. How should it interpret this? Answer: it cannot, because it is not the right audience for it.

**Review by kianenigma — CHANGES_REQUESTED** [review]
I strongly think usage of `TryStateSelect` is not correct here.

**Comment by Ank4n** [comment]
bot rebase

**Comment by Ank4n** [comment]
> Hey, is anyone still working on this? Due to the inactivity this issue has been automatically marked as stale. It will be closed if no further activity occurs. Thank you for your contributions.

Yes, will be reworking on this.
